### PR TITLE
refactor: Use IndexSet instead of Vec for collections of unique elements

### PR DESCRIPTION
**Pull request by olivier-lacroix**
This PR stores Pypi Dependencies, Dependencies, Platforms and Channels in IndexSet, instead of Vec.

There were a couple of TODOs in the code to that effect.

### CODE REVIEW

**Review by baszalmstra — COMMENTED** [review]
Im wondering what motivates this change?

Note that the order can be very important for requirements and therefor using a hashset might cause non-deterministic behavior. I think an IndexSet would be a better choice in these cases.

We used Vec in the manifest part of the code to remain as true to the input as possible, to not accidentally lose ordering information that is present in there.

**Comment by olivier-lacroix** [comment]
@baszalmstra there were TODOs in the code to that effect :-)

Noted re. IndexSet.

File: src/cli/project/platform/remove.rs [hunk 1]
```diff
@@ -38,7 +38,7 @@ pub async fn execute(mut project: Project, args: Args) -> miette::Result<()> {
     // Remove the platform(s) from the manifest
     project
         .manifest
-        .remove_platforms(&platforms, &feature_name)?;
+        .remove_platforms(platforms.clone(), &feature_name)?;
```
Comment (olivier-lacroix) [inline src/cli/project/platform/remove.rs:41]: Current (or previous) implementation does not notify the user is a platform cannot be found for the feature.

Is this desired? this seems like a different behaviour than for dependency removal, which will Err if the dependency is not found.

Comment (ruben-arts) [reply src/cli/project/platform/remove.rs:41]: Yeah that seems wrong, not removing without an error can lead to users thinking it removed something but they had a typo. Resulting in unexpected behavior.

**Comment by tdejager** [comment]
Other than the merge conflicts and maybe the fix that both @olivier-lacroix and @ruben-arts suggested. It looks good to me!

File: src/cli/project/platform/remove.rs (hunk 1 above)
Comment (olivier-lacroix) [reply src/cli/project/platform/remove.rs:41]: OK. I created an issue out of that, that can be implemented in a separate PR.

**Comment by olivier-lacroix** [comment]
Thanks @tdejager . Just rebased, and created an issue about the 'silent removal' behaviour. it can be tackled in a separate (and small) PR.

**Review by ruben-arts — APPROVED** [review]

**Comment by tdejager** [comment]
> Thanks @tdejager . Just rebased, and created an issue about the 'silent removal' behaviour. it can be tackled in a separate (and small) PR.

Excellent!

**Pull request merged**